### ISSUE TITLE
New Rule: Focus-Hover-Contrast

### ISSUE THREAD

**Issue by jkodu** [dequelabs/axe-core#1296]
# focus-hover-contrast

The rule tests color contrast between the text and background color for focusable elements, when in focus/ hover states. This is experimental rule.

**Tags:** wcag20, wcag143, wcag146

## Matcher

Any focusable element.

## Checks

Check focusable elements that have focus/hover styles from CSSOM, and compare color contrast of the text against the background (including any pseudo styles).

Return true - if color contrast is above desired level.
Return false - if color contrast is lower than desired level.
Return needs review - if contrast ratio cannot be determined.

Related Issue(s):
- #809

**Comment by jkodu** [comment]
Notes:

Fail Messages:
Fix all of the following (2 checks)
1. contrast when focused should be more than ...
2. contrast when hovered should be more than ...

**Comment by jkodu** [comment]
Further discussion Notes:

# focusable-text-contrast

The rule tests color contrast between the text and background color for focusable elements, when in focus/ hover states. This is experimental rule.

**Tags:** wcag20, wcag143, experimental

## Matcher

- Add `context` to matches. (Sep PR).
- Then filter focus/ hover styles from CSSOM, ensure child elements are not included if parent element has focus/ hover style.

Discussion Eg:
<body>
	<div class='one'>
		<a class='two'></a>
	</div>
<body>

body { background: yellow; }
.one:hover { background: red; }
.two:hover { color: white; }

## Checks

### Check 1
Compare color contrast of the text against the background (including any pseudo styles).

**Return true:** if color contrast is above desired level.
**Return false:** if color contrast is lower than desired level.
**Return needs review:** if contrast ratio cannot be determined.

### Check 2

**Comment by jkodu** [comment]
Note:

When the page does not have `cssom` for some reason, it is worth returning the root element for review with a note, that this page needs review.
